dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -619,12 +630,15 @@ void InterpreterSelectQuery::executeSingleQuery()
 			/** Оптимизация - если источников несколько и есть LIMIT, то сначала применим предварительный LIMIT,
 				* ограничивающий число записей в каждом до offset + limit.
 				*/
-			if (query.limit_length && hasMoreThanOneStream() && !query.distinct)
+			if (query.limit_length && hasMoreThanOneStream() && !query.distinct && !query.limit_by_expression_list)
 				executePreLimit();
 
 			if (need_second_distinct_pass)
 				union_within_single_query = true;
 
+			if (query.limit_by_expression_list && hasMoreThanOneStream())
```

Comment:
We should merge all streams together or condition streams.size() == 1 on line 645 will be false and executeLimit won't be called.